### PR TITLE
(DO NOT MERGE)(MODULES-2243) Remove LCM RefreshMode Change

### PR DESCRIPTION
**Pull request by jpogran**
Removes setting the LCM RefreshMode in the Invoke DSC Resource script and
have it bail out if it is set to anything else than Disabled.

Creates a Puppet type that will set the LCM RefreshMode to one of the
supported values

### CODE REVIEW

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb [hunk 1]
```diff
@@ -4,33 +4,10 @@ if (!(Test-Path($PuppetModulesFolder))) {
...
-
-  Set-DscLocalConfigurationManager -Path $configOutputPath
-}
-
 $currentState = Get-DscLocalConfigurationManager
 
 if ($currentState.RefreshMode -ne 'Disabled') {
-  $currentRefreshMode = $currentState.RefreshMode
-  Set-RefreshMode 'Disabled'
+  throw "DSC LCM RefreshMode must be set to Disabled for Puppet to execute DSC Resources!"
```
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: Perhaps not something to address now, but I wonder if we can better propagate error messages if we return something structured to Ruby upon error.

Food for thought.

File: templates/set-lcm-mode.ps1.erb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+Write-Host "Setting LCM"
+$LCMRefreshMode = '<%= @refresh_mode %>'
+$generationDate = (get-date -format "M/d/yyyy HH:mm:ss")
+$encoding       = [system.Text.Encoding]::UTF8
+$Namespace      = 'root/Microsoft/Windows/DesiredStateConfiguration'
+$ClassName      = 'MSFT_DSCLocalConfigurationManager'
+
+Write-Verbose "LCM Mode: $LCMRefreshMode"
```
Comment (Iristyle) [inline templates/set-lcm-mode.ps1.erb]: So FYI I don't think we have a way to pass in a `-Verbose` to PowerShell from Ruby...

File: templates/set-lcm-mode.ps1.erb [hunk 3]
```diff
@@ -0,0 +1,51 @@
+Write-Host "Setting LCM"
+$LCMRefreshMode = '<%= @refresh_mode %>'
+$generationDate = (get-date -format "M/d/yyyy HH:mm:ss")
+$encoding       = [system.Text.Encoding]::UTF8
```
Comment (Iristyle) [inline templates/set-lcm-mode.ps1.erb]: style nit - I usually omit the `system.` since it's not necessary

File: templates/set-lcm-mode.ps1.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+Write-Host "Setting LCM"
```
Comment (Iristyle) [inline templates/set-lcm-mode.ps1.erb]: style nit - Single quotes `'` vs double quotes `"`

File: templates/set-lcm-mode.ps1.erb (hunk 2 above)
Comment (jpogran) [inline templates/set-lcm-mode.ps1.erb]: The param set at line 42 sets verbose for the Invoke-CimMethod at line 51, so verbose messages show up. Line 1 should be a verbose message as well instead of write-host, but both were added during debugging and really don't add much value otherwise so it may be best just to remove both.

File: templates/set-lcm-mode.ps1.erb (hunk 3 above)
Comment (jpogran) [inline templates/set-lcm-mode.ps1.erb]: Really? I have always followed shortened on the command line, fully qualified (full namespace, no aliases, etc...) in the script.

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb (hunk 1 above)
Comment (jpogran) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: I don't think we're going to be able to achieve that without doing the json passing we have been discussing.

File: templates/set-lcm-mode.ps1.erb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+$ClassName      = 'MSFT_DSCLocalConfigurationManager'
+
+Write-Verbose "LCM Mode: $LCMRefreshMode"
+
+$mofData = @"
+/*
+@TargetNode=''localhost''
+@GeneratedBy=PUPPET
+@GenerationDate=$generationDate
+@GenerationHost=PUPPET
```
Comment (jpogran) [inline templates/set-lcm-mode.ps1.erb]: Was wondering if we should be setting this to the target node's hostname instead of Puppet since it's technically generated on the target node and not on puppet.

Comment (Iristyle) [inline templates/set-lcm-mode.ps1.erb]: Yeah, I think so.  Not sure if that value get's written into any extended logging anywhere, but seems like a good idea.

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb (hunk 1 above)
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: Yeah, we should probably file a ticket for that, and take care of it next sprint.

File: metadata.json [hunk 6]
```diff
@@ -33,10 +33,17 @@
     },
     {
       "name": "puppet",
-      "version_requirement": "3.x"
+      "version_requirement": ">= 3.4.0"
```
Comment (Iristyle) [inline metadata.json]: I would keep this specific line in another commit for bookkeeping purposes.

File: templates/set-lcm-mode.ps1.erb (hunk 3 above)
Comment (Iristyle) [inline templates/set-lcm-mode.ps1.erb]: Yeah, I tend to write code that same way - long form in scripts, except for a few exceptions.

* Omit `System.` since it's redundant and typically implied (which can help keep line length down)
* `%` instead of `ForEach-Object`
* `?` instead of `Where-Object`

I have a little style guide I wrote years ago around here somewhere.....

File: templates/set-lcm-mode.ps1.erb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  Verbose    = $true
+  CimClass   = $cimClass
+  CimSession = $wsmanSession
+  MethodName = "SendMetaConfigurationApply"
+  Arguments  = @{
+    ConfigurationData = $bytes;
+  }
+}
+
+Invoke-CimMethod @invokeParams
```
Comment (cowofevil) [inline templates/set-lcm-mode.ps1.erb]: Missing newline at end of file.

File: manifests/lcm_config.pp [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+) {
+
+  validate_re($refresh_mode, '^(Disabled|Pull|Push)$', 'refresh_mode must be one of \'Disabled\', \'Push\', \'Pull\'')
+
+  exec { "dsc_provider_set_lcm_refreshmode_${refresh_mode}":
+    provider => 'powershell',
+    onlyif   => "if ((Get-DscLocalConfigurationManager).RefreshMode -eq '${refresh_mode}') {exit 1} else {exit 0}",
+    command  => template('dsc/set-lcm-mode.ps1.erb'),
+  }
+}
```
Comment (cowofevil) [inline manifests/lcm_config.pp]: Missing newline at end of file.

File: metadata.json (hunk 6 above)
Comment (ferventcoder) [inline metadata.json]: That was my fault.

Comment (ferventcoder) [inline metadata.json]: @jpogran and it will be 3.7.0 by the way :)

File: metadata.json [hunk 9]
```diff
@@ -29,14 +29,17 @@
   "requirements": [
     {
       "name": "pe",
-      "version_requirement": "3.x"
-    },
-    {
-      "name": "puppet",
-      "version_requirement": "3.x"
+      "version_requirement": ">= 3.4.0"
```
Comment (ferventcoder) [inline metadata.json]: Pull this out completely. This will be best done with a different ticket.

Comment (ferventcoder) [inline metadata.json]: Plus it looks like it might have gotten a bit heavy handed and removed the puppet requirement.

**Comment by Iristyle** [comment]
In 9dd238f we probably want a `.fixtures.yml` as well to get those deps installed with `rake spec_prep`

```yml
fixtures:
  forge_modules:
    powershell: "puppetlabs/powershell"
    stdlib: "puppetlabs/stdlib"
  symlinks:
    dsc: "#{source_dir}"
```

File: manifests/lcm_config.pp [hunk 10]
```diff
@@ -0,0 +1,12 @@
+define dsc::lcm_config (
+  $refresh_mode = 'Disabled'
+) {
+
+  validate_re($refresh_mode, '^(Disabled|Pull|Push)$', 'refresh_mode must be one of \'Disabled\', \'Push\', \'Pull\'')
```
Comment (cowofevil) [inline manifests/lcm_config.pp]: Not to quibble, but the regex is case sensitive and Windows people are use to insensitive casing. May want to expect upper and lower case.

Comment (ferventcoder) [inline manifests/lcm_config.pp]: We haven't decided on that one yet. It depends on whether you can set the mode case insensitively or not.

File: .fixtures.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+fixtures:
+  forge_modules:
+    stdlib: "puppetlabs/stdlib"
+    powershell: "puppetlabs/powershell"
+  symlinks:
+    "dsc": "#{source_dir}"
```
Comment (Iristyle) [inline .fixtures.yml]: Missing newline ;0

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb (hunk 1 above)
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: I have confirmed that we probably want to improve this error reporting, because the way it bails right now is pretty user-unfriendly.  Certainly good enough for this PR, but we have room for improvement in a future iteration.


```
Error: /Stage[main]/Main/Dsc_registry[registry_test]: Could not evaluate: Execution of 'C:\Windows\sysnative\WindowsPowershell\v1.0\powershell.exe -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command $script:ErrorActionPreference = 'Stop'
$PuppetModulesFolder = Join-Path $env:ProgramFiles 'WindowsPowerShell\Modules\PuppetVendoredModules'
if (!(Test-Path($PuppetModulesFolder))) {
  & cmd.exe /c mklink /d "$PuppetModulesFolder" "c:\source\dsc\puppetlabs-dsc\spec\fixtures\modules\dsc\vendor\DSCResources" | Out-Null
}

$currentState = Get-DscLocalConfigurationManager

if ($currentState.RefreshMode -ne 'Disabled') {
  throw "DSC LCM RefreshMode must be set to Disabled for Puppet to execute DSC Resources!"
}

$invokeParams = @{
  Name          = 'Registry'
  Method        = 'test'
  ErrorAction   = 'Continue'
  ErrorVariable = 'invokeError'
  Property      = @{
    key = 'HKEY_LOCAL_MACHINE\SOFTWARE\TestKey'
    valuename = 'TestStringValue'
    valuedata = @('Cats')
    valuetype = 'String'
    ensure = 'Present'
  }
  ModuleName = "PSDesiredStateConfiguration"
}

$result = Invoke-DscResource @invokeParams

if ($currentRefreshMode) { Set-RefreshMode $currentRefreshMode }
if ($invokeError) { throw $invokeError }

switch ($invokeParams.Method) {
  'Test' {
    [bool]($result.InDesiredState)
  }
  'Set' {
    @($Null, 0) -contains $result.ReturnValue
  }
}
' returned 1: DSC LCM RefreshMode must be set to Disabled for Puppet to execute DSC Resources!
At line:10 char:3
+   throw "DSC LCM RefreshMode must be set to Disabled for Puppet to ex ...
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (DSC LCM Refresh... DSC Resources!:String) [], R
   untimeException
    + FullyQualifiedErrorId : DSC LCM RefreshMode must be set to Disabled for Puppet to execute
   DSC Resources!
```

**Comment by Iristyle** [comment]
Overall this is really close to merge as of e17e680.  

I'm pretty satisfied with local tests with multiple resource manifests, like this one:

```puppet
dsc_service { 'w32time':
  dsc_name  => 'w32time',
  dsc_state => 'Running'
}

dsc_registry { 'registry_test_binary':
  dsc_ensure    => 'Present',
  dsc_key       => 'HKEY_LOCAL_MACHINE\SOFTWARE\TestKey',
  dsc_valuename => 'TestBinaryValue',
  dsc_valuedata => 'BEEF',
  dsc_valuetype => 'Binary',
}

dsc_registry { 'registry_test_dword':
  dsc_ensure    => 'Present',
  dsc_key       => 'HKEY_LOCAL_MACHINE\SOFTWARE\TestKey',
  dsc_valuename => 'TestDwordValue',
  dsc_valuedata => '1',
  dsc_valuetype => 'Dword',
}

dsc_registry { 'registry_test_string':
  dsc_ensure    => 'Present',
  dsc_key       => 'HKEY_LOCAL_MACHINE\SOFTWARE\TestKey',
  dsc_valuename => 'TestStringValue',
  dsc_valuedata => 'Cats',
  dsc_valuetype => 'String',
}

dsc_service { 'w32time2':
  dsc_name  => 'w32time',
  dsc_state => 'Stopped'
}
```

Let's get #38 in first since we're going to want the rake task `spec_prep` working properly on Windows.  Then I believe you have a WIP test to update this PR with @jpogran, and some minor cleanup based on comments, and I think we can :shipit:
